components/s3_bucket/main.tf

```diff
@@ -62,6 +68,26 @@ resource "aws_s3_bucket" "bucket" {
     enabled = var.versioning || var.replication_target != null
   }
 
+  dynamic "lifecycle_rule" {
+    for_each = var.temporary_paths
+
+    content {
+      id      = md5(lifecycle_rule.value)
+      prefix  = lifecycle_rule.value
```

Comment:
What does the `prefix` do? Where does this value come from? Sorry still not super used to some of this syntax!

Reply:
No worries, it's a little hard to follow!! The `prefix` argument "filters" which objects in the S3 bucket this rule applies to (so by default, it'd apply to everything & this argument limits it to only things that match a given file/folder prefix). The `lifecycle_rule.value` is the "iterator" for this dynamic block & comes from the `var.lifecycle_rule` input variable that we're iterating over.

Reply:
So, in this case for Northstar, we're iterating over `["temporary/"]` and so we'd loop a single time where `lifecycle_rule.value` would be `"temporary/"` (and thus we'd only enforce these automatic deletion rules for files that were inside that folder).

Reply:
So in the case of Northstar this limits to stuff in `temporary/`?

Reply:
Yup! Here's what this turns into (from the plan) after Terraform does all it's fancy looping:

```hcl
lifecycle_rule {
  enabled = true
  prefix  = "temporary/"

  expiration {
    days = 14
  }

  noncurrent_version_expiration {
    days = 90
  }
}
```

Reply:
Ideally we'd be able to just create this as a standalone `aws_s3_lifecycle_rule` resource in the Northstar module that looked exactly like that & deal with any looping funny-business to build this into the `s3_bucket` module but alas! 🙃

Reply:
SO HELPFUL!! Thanks for spelling it out much much appreciated! I follow now!